nextest-filtering/src/parsing.rs

```diff
@@ -0,0 +1,668 @@
+    }
+
+    #[cfg(test)]
+    fn all() -> Expr {
+        Expr::Set(SetDef::All)
+    }
+
+    #[cfg(test)]
+    fn none() -> Expr {
+        Expr::Set(SetDef::None)
+    }
+}
+
+enum SpanLength {
+    Unknown,
+    Exact(usize),
+}
+
+fn expect_inner<'a, F, T>(
+    mut parser: F,
+    make_err: fn(SourceSpan) -> Error,
+    limit: SpanLength,
+) -> impl FnMut(Span<'a>) -> IResult<'a, Option<T>>
+where
+    F: FnMut(Span<'a>) -> IResult<T>,
+{
+    move |input| match parser(input) {
+        Ok((remaining, out)) => Ok((remaining, Some(out))),
+        Err(nom::Err::Error(err)) | Err(nom::Err::Failure(err)) => {
+            let nom::error::Error { input, .. } = err;
+            let start = input.location_offset();
+            let len = input.fragment().len();
+            let span = match limit {
+                SpanLength::Unknown => (start, len).into(),
+                SpanLength::Exact(x) => (start, x.min(len)).into(),
+            };
+            let err = make_err(span);
+            input.extra.report_error(err);
+            Ok((input, None))
+        }
+        Err(err) => Err(err),
+    }
+}
+
+fn expect<'a, F, T>(
+    parser: F,
+    make_err: fn(SourceSpan) -> Error,
+) -> impl FnMut(Span<'a>) -> IResult<'a, Option<T>>
+where
+    F: FnMut(Span<'a>) -> IResult<T>,
+{
+    expect_inner(parser, make_err, SpanLength::Unknown)
+}
+
+fn expect_char<'a>(
+    c: char,
+    make_err: fn(SourceSpan) -> Error,
+) -> impl FnMut(Span<'a>) -> IResult<'a, Option<char>> {
+    expect_inner(ws(char(c)), make_err, SpanLength::Exact(0))
+}
+
+fn silent_expect<'a, F, T>(mut parser: F) -> impl FnMut(Span<'a>) -> IResult<Option<T>>
+where
+    F: FnMut(Span<'a>) -> IResult<T>,
+{
+    move |input| match parser(input) {
+        Ok((remaining, out)) => Ok((remaining, Some(out))),
+        Err(nom::Err::Error(err)) | Err(nom::Err::Failure(err)) => {
+            let nom::error::Error { input, .. } = err;
+            Ok((input, None))
+        }
+        Err(err) => Err(err),
+    }
+}
+
+fn ws<'a, T, P: FnMut(Span<'a>) -> IResult<'a, T>>(
+    mut inner: P,
+) -> impl FnMut(Span<'a>) -> IResult<'a, T> {
+    move |input| {
+        let (i, _) = many0(char(' '))(input.clone())?;
+        match inner(i) {
+            Ok(res) => Ok(res),
+            Err(nom::Err::Error(err)) => {
+                let nom::error::Error { code, .. } = err;
+                Err(nom::Err::Error(nom::error::Error { input, code }))
+            }
+            Err(nom::Err::Failure(err)) => {
+                let nom::error::Error { code, .. } = err;
+                Err(nom::Err::Failure(nom::error::Error { input, code }))
+            }
+            Err(err) => Err(err),
+        }
+    }
+}
+
+fn is_identifier_char(c: char) -> bool {
+    // This is use for NameMatcher::Contains(_) and NameMatcher::Equal(_)
+    // The output should be valid part of a test-name or a package name.
+    c == ':' || c.is_xid_continue()
```

Comment:
Hmm, unfortunately this doesn't seem quite correct -- package names (and really test names with custom test harnesses) can have `-` in them, which is not XID_CONTINUE. In fact, test names can even be things like `foo.rs (line 30)` -- this is what doctests look like, for example.

I think we may want to accept arbitrary Unicode here for tests, and add ways to escape `(` and other special characters.